GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md

```diff
@@ -0,0 +1,420 @@
+    # type is appropriate
+    Operator::Function
+end
+
+## example for other computation models
+type AdiaComput <: QuComputModel
+    Operator::Function
+    Hb::AbstractMatrix
+    Hp::AbstractMatrix
+end
+```
+
+### At the end of the project, the result would look as follows:
+
+Prototype 1: Simple Quantum Circuits
+## Construct a circuit
+a circuit is constructed by `Circuit`:
+
+you can construct an empty circuit:
+```julia
+circuit = Circuit()
+```
+
+to add gates, use the method `addgate!`
+
+the `addgate!` method allows following way to add a quantum gate or a module to a specific place in the circuit.
+
+```julia
+# for single qubit gates(like Hadamard)
+# circuit : the circuit object
+# gate : gate object you want to insert
+# id   : the id of the qubit you want to add this gate to
+addgate!(circuit,gate,id)
+
+# for example
+# |a> ----[H]----
+# |b> -----------
+# |c> -----------
+# let's add a Hadamard gate to the first qubit in this three-qubit circuit
+
+# first create your circuit
+c = Circuit()
+addgate!(c,Hadamard,1)
+
+# for controlled gates
+# |a> -----x--------
+#          |
+# |b> ----[H]-------
+# |c> --------------
+
+c_H = Circuit()
+addgate!(c_H, 1, Hadamard, 2)
+# or for multiple input gates(will be useful for other gate-like modules)
+
+# |a> --------x------
+#             |
+# |b> ------|***|----
+#           | M |
+# |c> ------|***|----
+c_M = Circuit()
+addgate!(c_M, 1, M, 2, 3)
+
+# or for multiple controlled gates
+
+# |a> -------x--------
+#            |
+# |b> -------x--------
+#            |
+# |c> ------[H]-------
+cc_H = Circuit()
+addgate!(cc_H, [1,2], Hadamard, 3)
+
+```
+
+to measure a circuit, simply use the `measure` method on the circuit you want to measure with the the circuit's input state
+
+```julia
+# measure the circuit:
+# |a> ----[H]----
+# |b> -----------
+# |c> -----------
+# with input state:
+# 1/2*sqrt(2)(|000>+|001>+|010>+...+|111>)
+
+input_state = InitState(3)
+circuit = Circuit()
+addgate!(circuit, Hadamard, 1)
+res_state = measure(circuit,input_state)
+```
+
+Remove a gate from circuit
+```julia
+# given a circuit
+# |a> ----[H]-------
+# |b> ---------[H]--
+# |c> --------------
+
+removegate!(circuit,1)
+
+# remove the first gate
+# |a> --------------
+# |b> --------[H]---
+# |c> --------------
+```
+
+Prototype 2: quantum  Fourier Transformation
+![QFT circuit pic](imag/fourier.svg)
+
+```julia
+# Quantum Fourier Transformation
+init_state = InitStates(4)# inputs the number of bits
+
+QFT_curcuit = Circuit()
+
+for i=1:4
+    addgate!(QFT_circuit,Hadamard,i)
+    for j=1:i-1
+        addgate!(QFT_circuit,bitID-i,R_k(i),1)
+    end
+end
+
+res_state = measure(QFT_circuit,init_state)
+# suppose this to be the command line(the result is made up)
+---
+measurement of QFT_circuit:
+states| prob
+|0000>: 0.0001
+|0001>: 0.2000
+...
+|1111>: 0.2000
+---
+
+figure(1)
+# For visualization, a circuit should be plotted by
+plot(QFT_circuit)
+show()
+```
```

Comment:
A reference should do but if you can manage to work a prototype it would be nice as well :)